cirq/experiments/cross_entropy_benchmarking.py

```diff
@@ -0,0 +1,347 @@
+        """
+        Args:
+            num_cycle_range: The different numbers of cycles (circuit depth) at
+                which the measurements were taken.
+            xeb_fidelities: The average XEB fidelity at each number of cycles.
+        """
+        self._num_cycles = num_cycle_range
+        self._xeb_fidelities = xeb_fidelities
+
+    @property
+    def data(self) -> Sequence[Tuple[int, float]]:
+        """Returns a sequence of tuple pairs with the first item being a
+        number of cycles and the second item being the corresponding average
+        XEB fidelity.
+        """
+        return [(num, prob) for num, prob in zip(self._num_cycles,
+                                                 self._xeb_fidelities)]
+
+    def plot(self, **plot_kwargs: Any) -> None:
+        """Plots the average XEB fidelity vs the number of cycles.
+
+        Args:
+            **plot_kwargs: Arguments to be passed to matplotlib.pyplot.plot.
+        """
+        fig = plt.figure()
+        ax = plt.gca()
+        ax.set_ylim([0, 1.1])
+
+        plt.plot(self._num_cycles, self._xeb_fidelities, 'ro-',
+                 figure=fig, **plot_kwargs)
+        plt.xlabel('Number of Cycles', figure=fig)
+        plt.ylabel('XEB Fidelity', figure=fig)
+        fig.show()
+
+
+def cross_entropy_benchmarking(
+        sampler: sim.Sampler,
+        qubits: Sequence[devices.GridQubit],
+        two_qubit_gate: ops.gate_features.TwoQubitGate,
+        *,
+        num_circuits: int = 20,
+        repetitions: int = 1000,
+        num_cycle_range: Sequence[int] = range(2, 103, 10),
+        interaction_sequence: List[Set[Tuple[devices.GridQubit,
+                                             devices.GridQubit]]] = None,
+        use_tetrahedral_group: bool = False) -> CrossEntropyResult:
+    r"""Cross-entropy benchmarking (XEB) of multiple qubits.
+
+    A total of M random circuits are generated, each of which is made of N
+    cycles where N = max(num_cycle_range). Every cycle contains randomly
+    generated single-qubit gates applied to each qubit, followed by two-qubit
+    gate(s) applied to one or more pairs of qubits. For each given number of
+    cycles n in num_cycle_range, the experiment performs the following:
+
+    1) Measure the bit-string probabilities P^(meas, m)_|...00>,
+    P^(meas, m)_|...01>, P^(meas, m)_|...10>, P^(meas, m)_|...11>... at the
+    end of circuit_mn for all m, where circuit_mn is the first n cycles of
+    random circuit m (m <= M).
+
+    2) Theoretically compute the expected bit-string probabilities
+    P^(exp, m)_|...00>,  P^(exp, m)_|...01>, P^(exp, m)_|...10>,
+    P^(exp, m)_|...11>... at the end of circuit_mn for all m.
+
+    3) Collect all measured probabilities into a single vector P_meas and
+    normalize P_meas such it sums to 1.
+
+    4) Collect all expected probabilities into a single vector P_exp. Set any
+    probabilities <1e-22 to 1e-22 to avoid singularities in log P_exp. Then
```

Comment:
This is equivalent to omitting values that you didn't sample.

Equivalently, you could think of it as using Laplace's rule of succession. In the binary case your estimated probability is never 0/N, it's at least 1/(N+1).

Reply:
`P_exp` actually refers to the "expected" or theoretically computed probabilities here (`P_meas` is the experimental probability). I think setting `P_exp` to some arbitrarily small value is mostly to avoid the log throwing an error. We could have set it to 1e-25 or 1e-20 or any other really small number without affecting the value of the XEB fidelity.

Reply:
Could you double-check with Sergio whether we still compute the logarithms of the probabilities? If we didn't, this would be an unnecessary value.

I don't like setting an arbitrary threshold like this in a statistical computation. If we're going to omit samples that didn't occur, we should just omit samples that didn't occur instead of doing it in a roundabout way.